### PR TITLE
fix(client): avoid calling the api for signed out users

### PR DESCRIPTION
**Pull request by ahmaxed**
Users who are not singed in would get an "Access token is required for this request" error before this change.

### CODE REVIEW

File: client/src/components/Donation/PaypalButton.js [hunk 1]
```diff
@@ -41,8 +41,8 @@ export class PaypalButton extends Component {
     const { amount, duration } = this.state;
     const { skipAddDonation = false } = this.props;
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/PaypalButton.js:42]:
```suggestion
    const { isSignedIn = true } = this.props;
```

File: client/src/components/Donation/PaypalButton.js [hunk 2]
```diff
@@ -41,8 +41,8 @@ export class PaypalButton extends Component {
...
 
-    // Skip the api if user is not signed in or if its a one-time donation
-    if (!skipAddDonation || isSubscription) {
+    // Skip the api if user is not signed in and if it is a one-time donation
+    if (!skipAddDonation && isSubscription) {
```
Comment (ojeytonwilliams) [inline client/src/components/Donation/PaypalButton.js:45]:
```suggestion
    if (isSignedIn && isSubscription) {
    // call the api
```
Bit nitpicky, but the comment and the logic were in conflict.  Also, I think this is a fraction easier to understand.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Could we just us `isSignedIn` everywhere?  It doesn't seem to get us anything to use `skipAddDonation` since the logic isn't even just `if(skipAddDonation) { // skip donation...`

File: client/src/components/Donation/PaypalButton.js (hunk 2 above)
Comment (ShaunSHamilton) [reply client/src/components/Donation/PaypalButton.js:45]: I would also prefer this, as it is not immediately obvious when and why `skipAddDonation` is used.

File: client/src/components/Donation/PaypalButton.js (hunk 1 above)
Comment (ShaunSHamilton) [reply client/src/components/Donation/PaypalButton.js:42]: Should we be defaulting to `true`? I do not see it happening anywhere else in the codebase.

**Review by raisedadead — COMMENTED** [review]
Once you have addressed the review comments, some tests would lock this bug from appearing again.

File: client/src/components/Donation/PaypalButton.js (hunk 1 above)
Comment (ahmadabdolsaheb) [reply client/src/components/Donation/PaypalButton.js]: I just passed the isSignedIn to the modal as well, so thins change is not necessary.

File: client/src/components/Donation/PaypalButton.js (hunk 2 above)
Comment (ahmadabdolsaheb) [reply client/src/components/Donation/PaypalButton.js]: Done

**Review by ojeytonwilliams — DISMISSED** [review]
LGTM :+1: 

Pending tests, I think this is good to go.

**Comment by raisedadead** [comment]
@ahmadabdolsaheb ping.

**Comment by raisedadead** [comment]
Hi @ahmadabdolsaheb this PR is has been pending a merge for the need of tests and merge conflicts.

**Comment by ahmadabdolsaheb** [comment]
Thanks @raisedadead for the reminder 🙏. I am struggling to get cypress to make api calls locally ATM. I will make the changes and possibly pair with another member to finalize the tests.

**Comment by raisedadead** [comment]
Unit tests should be fine for this.

File: client/src/components/Donation/PaypalButton.test.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+const donationData = {
+  redirecting: false,
+  processing: false,
+  success: false,
+  error: null
+};
+
+interface Ref {
```
Comment (ShaunSHamilton) [inline client/src/components/Donation/PaypalButton.test.tsx:21]: You should not need to define a `Ref` interface. This should come directly from React: [Example](https://github.com/freeCodeCamp/freeCodeCamp/blob/main/client/src/components/Header/components/universal-nav.tsx#L7)

Comment (ShaunSHamilton) [reply client/src/components/Donation/PaypalButton.test.tsx:21]: Also, seeing as this is a new file, shall we change this to kebab-case?

Comment (ahmadabdolsaheb) [reply client/src/components/Donation/PaypalButton.test.tsx]: Thanks, both requested changes are added.

**Review by raisedadead — COMMENTED** [review]
Both tests should not use the same test description and are currently lacking clarity. Also, there should be a test for the user story that this PR is aiming to fix in the first place, that is when they are not signed in.

**Comment by ahmadabdolsaheb** [comment]
Thanks for catching that. it should do it now.

**Review by raisedadead — APPROVED** [review]

**Review by ShaunSHamilton — APPROVED** [review]
Approval without merge. Merge if happy, @raisedadead 

I am happy with the logic changes. However, after multiple attempts, I have been unable to test the donations. Would appreciate more assistance with how to QA the donation stuff.

**Review by moT01 — APPROVED** [review]
LGTM 🎉 

@ShaunSHamilton - there are test keys in our password tool. You can add them to your .env - make sure they're the test keys. Then, I use the [fake card info](https://stripe.com/docs/testing) here. An alternative would be to create your own dev accounts at paypal and stripe and use those keys. There may be some additional things you need to create if you want to go that route. I'm thinking of probably subscription ID's - but with it, you can see more of what's happening on their ends.

**Pull request merged**